File: parse_annotations.js
const fs = require('fs');
const path = require('path');

// 指向新的文件
const INPUT_FILE = "d:\\桌面2.0\\id598.md";
const OUTPUT_FILE = path.join(__dirname, "lib", "mock-data-final.ts");

function parseJson() {
    try {
        let rawData = fs.readFileSync(INPUT_FILE, 'utf-8');
        
        // 提取 Markdown 代码块中的 JSON
        const jsonMatch = rawData.match(/```\s*(\{[\s\S]*\})\s*```/);
        if (jsonMatch) {
            rawData = jsonMatch[1];
        } else {
            // 尝试直接查找第一个 { 和 最后一个 }
            const start = rawData.indexOf('{');
            const end = rawData.lastIndexOf('}');
            if (start !== -1 && end !== -1) {
                rawData = rawData.substring(start, end + 1);
            }
        }

        const data = JSON.parse(rawData);
        
        // 我们只关心 id: 598 的数据
        // 这个文件本身就是一个 task 对象 (id: 598)
        let results = [];
        if (data.id === 598) {
             const annotations = data.annotations || [];
             if (annotations.length > 0) {
                 results = annotations[0].result || [];
             }
        } else {
            console.error("Warning: Task ID is not 598, but proceeding anyway.");
            const annotations = data.annotations || [];
             if (annotations.length > 0) {
                 results = annotations[0].result || [];
             }
        }

        const uniqueBoxes = []; 
        const THRESHOLD = 1.0; 

        results.forEach((item, index) => {
            if (item.type !== 'rectanglelabels') return;
            
            const val = item.value || {};
            const labels = val.rectanglelabels || [];
            
            let glomType = "cellular";
            let priority = 1;

            const labelStr = labels.join(",");
            
            // 优先级逻辑
            if (labelStr.includes("硬化")) { glomType = "sclerotic"; priority = 10; }
            else if (labelStr.includes("新月体")) { glomType = "crescents"; priority = 9; }
            else if (labelStr.includes("系膜")) { glomType = "membranous"; priority = 8; }
            else if (labelStr.includes("固有")) { glomType = "normal"; priority = 7; }
            else if (labelStr.includes("毛细血管")) { glomType = "cellular"; priority = 5; }

            const cx = val.x + val.width / 2;
            const cy = val.y + val.height / 2;

            let existingIndex = -1;
            for (let i = 0; i < uniqueBoxes.length; i++) {
                const box = uniqueBoxes[i];
                const boxCx = box.x + box.width / 2;
                const boxCy = box.y + box.height / 2;
                const dist = Math.sqrt(Math.pow(cx - boxCx, 2) + Math.pow(cy - boxCy, 2));
                if (dist < THRESHOLD) {
                    existingIndex = i;
                    break;
                }
            }

            const newBox = {
                x: val.x,
                y: val.y,
                width: val.width,
                height: val.height,
                type: glomType,
                priority: priority,
                confidence: item.score || 0.95
            };

            if (existingIndex !== -1) {
                // 如果是“重复框”，我们取信息量最大的那个（或者最后那个）
                if (newBox.priority > uniqueBoxes[existingIndex].priority) {
                    uniqueBoxes[existingIndex] = newBox;
                }
            } else {
                uniqueBoxes.push(newBox);
            }
        });
            
        // 写入文件
        let fileContent = "// 自动生成的数据，请手动复制到 web/lib/mock-data.ts\n";
        fileContent += `// Source: ID 598\n`;
        fileContent += `// Total Unique Count: ${uniqueBoxes.length}\n`;
        fileContent += "export const DATA_FINAL_IMG = [\n";
        
        uniqueBoxes.forEach((item, index) => {
            fileContent += `  { id: 'final-${index}', x: ${item.x.toFixed(4)}, y: ${item.y.toFixed(4)}, width: ${item.width.toFixed(4)}, height: ${item.height.toFixed(4)}, type: '${item.type}', confidence: ${item.confidence.toFixed(2)} },\n`;
        });
        
        fileContent += "];\n";
        
        fs.writeFileSync(OUTPUT_FILE, fileContent, 'utf-8');
        console.log(`Successfully wrote ${uniqueBoxes.length} annotations to ${OUTPUT_FILE}`);

    } catch (e) {
        console.error("Error processing file:", e);
    }
}

parseJson();

File: parse_annotations.py
import json
import sys

# 设置输入输出文件路径
INPUT_FILE = r"d:\桌面2.0\project-16-at-2025-12-11-11-32-738730a2.json"

# 类型映射表
TYPE_MAP = {
    "硬化性肾小球": "sclerotic",
    "硬化肾小球": "sclerotic",
    "新月体肾小球": "crescents", 
    "新月体性肾小球": "crescents",
    "伴有新月体形成的肾小球": "crescents",
    "新月体形成的肾小球": "crescents",
    "单纯系膜增生性肾小球": "membranous",
    "单纯系膜增生肾小球": "membranous",
    "毛细血管内增生性肾小球": "cellular",
    "毛细血管内增生肾小球": "cellular",
    "毛细血管细胞增生的肾小球": "cellular",
    "固有细胞无明显增生性肾小球": "normal",
    "肾小球": "cellular" # 默认归类，或者根据置信度/其他逻辑
}

def parse_json():
    try:
        with open(INPUT_FILE, 'r', encoding='utf-8') as f:
            data = json.load(f)
        
        # 假设 JSON 是 Label Studio 导出的格式，通常是一个列表，每一项代表一个任务
        # 我们只需要第一个任务的结果
        if isinstance(data, list) and len(data) > 0:
            annotations = data[0].get('annotations', [])
            if len(annotations) > 0:
                results = annotations[0].get('result', [])
            else:
                print("Error: No annotations found.")
                return
        else:
             # 有时候直接是 dict
             results = data.get('annotations', [{}])[0].get('result', [])

        ts_output = []
        
        for i, item in enumerate(results):
            if item.get('type') != 'rectanglelabels':
                continue
            
            val = item.get('value', {})
            labels = val.get('rectanglelabels', [])
            
            # 确定类型
            # 优先匹配具体的病理类型，如果没有则使用默认
            glom_type = "cellular" # 默认值
            
            for label in labels:
                if label in TYPE_MAP:
                    glom_type = TYPE_MAP[label]
                    break
                # 模糊匹配
                for k, v in TYPE_MAP.items():
                    if k in label:
                        glom_type = v
                        break
            
            # 构造对象
            obj = {
                "id": f"final-{i}",
                "x": val.get('x', 0),
                "y": val.get('y', 0),
                "width": val.get('width', 0),
                "height": val.get('height', 0),
                "type": glom_type,
                "confidence": 0.95 + (i % 5) * 0.01 # 模拟置信度
            }
            ts_output.append(obj)
            
        # 生成 TypeScript 代码
        print("const DATA_FINAL_IMG: Glomerulus[] = [")
        for item in ts_output:
            print(f"  {{ id: '{item['id']}', x: {item['x']:.4f}, y: {item['y']:.4f}, width: {item['width']:.4f}, height: {item['height']:.4f}, type: '{item['type']}', confidence: {item['confidence']:.2f} }},")
        print("];")
        
        # 打印所有遇到的标签，用于调试
        print("\n// Detected Labels:")
        all_labels = set()
        for item in results:
            if 'rectanglelabels' in item.get('value', {}):
                all_labels.update(item['value']['rectanglelabels'])
        print(f"// {all_labels}")

    except Exception as e:
        print(f"Error processing file: {e}")

if __name__ == "__main__":
    parse_json()

